sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java

```diff
@@ -235,49 +215,88 @@ public Size apply(
-                  invokeTargetMethod(instrumentedMethod));
+          // Figure out how many locals we'll need. This corresponds to "this", the parameters
+          // of the instrumented method, and an argument to hold the return value if the target
+          // method has a return value.
+          int numLocals = 1 + instrumentedMethod.getParameters().size() + (targetHasReturn ? 1 : 0);
+
+          Integer returnVarIndex = null;
+          if (targetHasReturn) {
+            // Local comes after formal parameters, so figure out where that is.
+            returnVarIndex = 1; // "this"
+            for (Type param : Type.getArgumentTypes(instrumentedMethod.getDescriptor())) {
+              returnVarIndex += param.getSize();
+            }
+          }
+
+          StackManipulation manipulation = new StackManipulation.Compound(
+              // Push "this" (DoFnInvoker on top of the stack)
+              MethodVariableAccess.REFERENCE.loadOffset(0),
+              // Access this.delegate (DoFn on top of the stack)
+              FieldAccess.forField(delegateField).getter(),
+              // Run the beforeDelegation manipulations.
+              // The arguments necessary to invoke the target are on top of the stack.
+              beforeDelegation(instrumentedMethod),
+              // Perform the method delegation.
+              // This will consume the arguments on top of the stack
+              // Either the stack is now empty (because the targetMethod returns void) or the
+              // stack contains the return value.
+              new UserCodeMethodInvocation(returnVarIndex, targetMethod, instrumentedMethod),
+              // Run the afterDelegation manipulations.
+              // Either the stack is now empty (because the instrumentedMethod returns void)
+              // or the stack contains the return value.
+              afterDelegation(instrumentedMethod));
+
           StackManipulation.Size size = manipulation.apply(methodVisitor, implementationContext);
-          return new Size(size.getMaximalSize(), instrumentedMethod.getStackSize());
+          return new Size(size.getMaximalSize(), numLocals);
         }
       };
     }
 
     /**
-     * Generates code to invoke the target method. When this is called the delegate field will be on
-     * top of the stack. This should add any necessary arguments to the stack and then perform the
-     * method invocation.
+     * Return the code to the prepare the operand stack for the method delegation.
+     *
+     * <p>Before this method is called, the stack delegate will be the only thing on the stack.
+     *
+     * <p>After this method is called, the stack contents should contain exactly the arguments
+     * necessary to invoke the target method.
      */
-    protected abstract StackManipulation invokeTargetMethod(MethodDescription instrumentedMethod);
+    protected abstract StackManipulation beforeDelegation(MethodDescription instrumentedMethod);
```

Comment:
Coming back to this after some time, I think it would aid readability to describe the difference between `targetMethod` and `instrumentedMethod`.